### PR TITLE
Semantic chunker #88

### PR DESCRIPTION
**Pull request by sambhavnoobcoder**
# Peak Detection Enhancement for SemanticChunker

## Problem Statement
The SemanticChunker class needed an enhancement to better identify and handle semantic boundaries in text by detecting peaks in semantic similarity scores. This would improve the quality of text chunking by more accurately identifying natural breakpoints in the content.

## Issue Analysis
The existing implementation of SemanticChunker had limitations in identifying optimal chunk boundaries, particularly when dealing with complex text structures. This could lead to chunks that either split semantically related content or combine unrelated content.

## Solution Design
We implemented a peak detection mechanism that:
1. Analyzes semantic similarity scores across text segments
2. Identifies significant peaks that represent natural semantic boundaries
3. Integrates with the existing chunking logic to create more meaningful chunks

## Implementation Details
The solution was implemented in multiple phases:

1. **Dependency Updates**
   - Updated necessary dependencies in `pyproject.toml` to support the new functionality

2. **Core Implementation**
   - Enhanced `semantic.py` with peak detection logic
   - Created a new `peak_detection.py` module with specialized algorithms
   - Modified the chunking process to incorporate peak detection

3. **Testing Infrastructure**
   - Added comprehensive test suite in `test_peak_detection.py`
   - Created integration tests in `test_semantic_chunker_peak_detection.py`
   - Ensured backward compatibility with existing functionality

## Testing Strategy
The implementation was thoroughly tested through:
1. Unit tests for peak detection algorithms
2. Integration tests for SemanticChunker with peak detection
3. Edge case handling for various text structures
4. Performance testing with different text lengths and complexities

## Screenshots/Demos
image

**fixes:** #88 

Lets get CHONKing !!!!!!!!!!

### CODE REVIEW

File: src/chonkie/chunker/peak_detection.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        """
+        # Find zero crossings in first derivative
+        zero_crossings = np.where(np.diff(np.signbit(first_deriv)))[0]
+        
+        if self.find == 'minima':
+            # Minima: positive second derivative
+            extrema = [
+                idx for idx in zero_crossings
+                if second_deriv[idx] > 0
+            ]
```
[the remark this answers is not in the document]
Comment (Sambhav-at-Atomicwork) [reply src/chonkie/chunker/peak_detection.py:117]: addressed in [d1cc4c8](https://github.com/chonkie-inc/chonkie/pull/115/commits/d1cc4c8c10e29f1e457a6b3dc27c04cac83ba587)

File: src/chonkie/chunker/peak_detection.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+                extrema = [
+                    idx for idx in extrema
+                    if signal[idx] <= self.threshold
+                ]
+        else:
+            # Maxima: negative second derivative
+            extrema = [
+                idx for idx in zero_crossings
+                if second_deriv[idx] < 0
+            ]
```
[the remark this answers is not in the document]
Comment (Sambhav-at-Atomicwork) [reply src/chonkie/chunker/peak_detection.py:133]: addressed in [b1031af](https://github.com/chonkie-inc/chonkie/pull/115/commits/b1031af1e24525b09596c9b2aec8132922a680c2)

File: src/chonkie/chunker/peak_detection.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sambhavnoobcoder) [reply src/chonkie/chunker/peak_detection.py:117]: addressed in [d1cc4c8](https://github.com/chonkie-inc/chonkie/pull/115/commits/d1cc4c8c10e29f1e457a6b3dc27c04cac83ba587)

File: src/chonkie/chunker/peak_detection.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sambhavnoobcoder) [reply src/chonkie/chunker/peak_detection.py:133]: addressed in [b1031af](https://github.com/chonkie-inc/chonkie/pull/115/commits/b1031af1e24525b09596c9b2aec8132922a680c2)

File: src/chonkie/chunker/semantic.py [hunk 3]
```diff
@@ -392,30 +411,63 @@ def _compute_window_similarities(
...
+        Returns:
+            List of indices where splits should occur
+        """
+        if self.use_peak_detection:
+            # Convert similarities to numpy array
+            similarities_array = np.array(similarities)
+            
+            # Detect peaks in the similarity signal
+            # Note: We invert the signal since we want to find minima
+            peaks = self.peak_detector.detect_peaks(1 - similarities_array)
```
[the remark this answers is not in the document]
Comment (sambhavnoobcoder) [reply src/chonkie/chunker/semantic.py:436]: addressed in [78fef00](https://github.com/chonkie-inc/chonkie/pull/115/commits/78fef001d011842113d4599da1d1b9d555002e7d)

**Comment by sambhavnoobcoder** [comment]
@chonknick i think this is ready for review , kindly give this a look .

**Comment by chonknick** [comment]
Hey @sambhavnoobcoder! 👋

WOAH! This is awesome~ Thanks for opening this PR, I've been thinking about having this for a while! 🚀

Just want to be super transparent here with you, this is something that I'm hoping to add as a breaking change~ so it would take some time for us to merge this as well (since we would get to a stable 1.0.x before moving with all the breaking changes). But this is definitely needed!

Some initial thoughts that I have are (haven't gone through the detailed code-review yet):

- Can we avoid the `scipy` dependency somehow? It's quite heavy and I would love if we can just have the smoothening filter + differentiator implemented from scratch with just numpy. 
- Could we have a benchmark comparison with the current implementation of the SemanticChunker to see if it improves or decreases performance? Possibly with a text like `enwik8` or `War and Peace`. 

Again, superb job, appreciate the effort on this!

Thanks! 😊
